### DISCUSSION TITLE
CIre unexpected behavior

### DISCUSSION THREAD

**Discussion by Antguz** [davidfrantz/force#194] [Q&A] [answered]
Hi,

I am estimating CIre using FORCE HLPS and had found unexpected behavior where high values seem to be truncated, and thus, its remove pixels with high chlorophyll content like forests or crops. 

To me this is a bug (please correct me) since is very odd that pixels from forest regions at different periods of time are removed for this index, but not NDVI for instance. See the image below for the CIre index in three periods during 2019.

![image](https://user-images.githubusercontent.com/7254767/170304106-57949187-1dbc-4571-a5e7-0ef1600e77bf.png)

I read that FORCE v. 3.7.5 was expected to solve that issue, and for that reason, I was running 3.7.6.

The histograms seem to reveal this truncated behavior at high values. 

![image](https://user-images.githubusercontent.com/7254767/170300486-0600c72e-ee4c-4649-aa3a-dcbb9457ba4d.png)

In the link enclosed below to a zip file, you will be able to reproduce and see this behavior using a title.

https://drive.google.com/file/d/1C5KO7lsVHi1dzt2zUx0KBJsUHcqJCBfw/view?usp=sharing

Please let me know if this is not correct.

AG.

**Comment by davidfrantz** [comment]
Hi @Antguz,

the fix was more on quantization: https://github.com/davidfrantz/force/issues/152

Yes, it seems like scaling with 10000 is too much. What would be the possible value range of CIre?

@danschef, do you know more?

Cheers,
David

**Reply by danschef** [reply]
Yes, I agree, the CIre is computed with:

(RedEdge3 / RedEdge1) -1

Thus, there is no strict upper limit and apparently it can happen that the index results in values above 3. Scaled by 10000, large CIre values can therefore exceed the upper bound of the signed integer 16bit data type (32.768), which is also visible in the histogram above.

Not sure how to solve that since all FORCE indices are produced as int16, scaled by 10000, right?

**Reply by Antguz** [reply]
Hi @davidfrantz and @danschef,

Is there any way that you can do a prior scaling of the value within the index calculation before the 10000 export scale? 

Let's says that you first compute the index like ((RedEdge3 / RedEdge1) -1) / 10, and the resulting value will be multiplied by 10000. That will give you values with a reasonable distribution between -10000 and 10000.

![Screenshot from 2022-05-26 13-27-12](https://user-images.githubusercontent.com/7254767/170552549-725647a3-080a-4e4c-9f38-8f1277c5669b.png)
 
It will be just a modification in a line of code, line 172 of `index-hl.c`. 

```
void index_ratio_minus1(ard_t *ard, small *mask_, tsa_t *ts, int b1, int b2, int nc, int nt, short nodata){
int p, t;
float ind, scale = 10000.0;

  #pragma omp parallel private(t,ind) shared(ard,mask_,ts,b1,b2,nc,nt,nodata,scale) default(none)
  {

    #pragma omp for
    for (p=0; p<nc; p++){

      if (mask_ != NULL && !mask_[p]){
        for (t=0; t<nt; t++) ts->tss_[t][p] = nodata;
        continue;
      }

      for (t=0; t<nt; t++){

        if (!ard[t].msk[p]){
          ts->tss_[t][p] = nodata;
        } else {
          ind = ((ard[t].dat[b1][p] / (float)ard[t].dat[b2][p]) - 1.0) / 10;
          if (ard[t].dat[b2][p] == 0 || ind*scale > SHRT_MAX || ind*scale < SHRT_MIN){
            ts->tss_[t][p] = nodata;
          } else {
            ts->tss_[t][p] = (short)(ind*scale);
          }
        }

      }

    }
  }

  return;
}
```

Then in the documentation you can highlight that this index is scaled at 1000 and not 10000. That is the easiest solution that comes to my mind.

AG.

**Comment by davidfrantz** [comment]
Hi @Antguz,

it would be simpler to just change the scaling factor in the 3rd line of the snippet that you posted. Will x1000 give you enough quantization?

David

**Reply by Antguz** [reply]
Hi @davidfrantz,

Of course that will be easier! I missed that line. 

I think that this truncated effect could happen in any simple ratio index; thus changing the scale to 1000 for index type could be beneficial for all of them.

AG.

**Comment by davidfrantz** [comment, marked as answer]
This is now fixed with https://github.com/davidfrantz/force/commit/f255a4bfa27a62ab592ef09c24ee8223b93dd565 in the develop branch.
Thanks AG!
